drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,454 @@
+#define RTC_NXP_RTCTIME_COMPARE_FLAG(mask, flag) (((mask) == 0) || (((mask) & (flag)) == (flag)))
+
+/* Utility function to check the bounds of passed in rtc_time */
+static uint8_t is_rtc_time_in_bounds(const struct rtc_time *time_value, uint16_t mask)
+{
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_SECOND)) &&
+	    (time_value->tm_sec < 0 || time_value->tm_sec > 59)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_MINUTE)) &&
+	    (time_value->tm_min < 0 || time_value->tm_min > 59)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_HOUR)) &&
+	    (time_value->tm_hour < 0 || time_value->tm_hour > 23)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_MONTHDAY)) &&
+	    (time_value->tm_mday < 1 || time_value->tm_mday > 31)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_MONTH)) &&
+	    (time_value->tm_mon < 0 || time_value->tm_mon > 11)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_YEAR)) &&
+	    ((time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) < -128 ||
+	     (time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) > 127)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_WEEKDAY)) &&
+	    (time_value->tm_wday < -1 || time_value->tm_wday > 6)) {
+		return 0;
+	}
+
+	if ((mask == 0) && (time_value->tm_isdst < -1 || time_value->tm_isdst > 1)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+/*
+ * The runtime where this is accessed is unknown so we force a lock on the registers then force an
+ * unlock to guarantee 2 seconds of write time
+ */
+static void nxp_irtc_unlock_registers(RTC_Type *reg)
+{
+	/* Lock the regsiters */
+	while ((reg->STATUS & (uint16_t)RTC_STATUS_WRITE_PROT_EN_MASK) == 0) {
+		*(uint8_t *)(&reg->STATUS) |= RTC_STATUS_WE(0x2);
+	}
+	/* Unlock the registers */
+	while ((reg->STATUS & (int16_t)RTC_STATUS_WRITE_PROT_EN_MASK) != 0) {
+		/*
+		 * The casting is required for writing only a single Byte to the STATUS Register,
+		 * the pattern here unlocks all RTC registers for writing. When unlocked if a 0x20
+		 * if written to the STATUS register the RTC registers will lock again and the next
+		 * write will lead to a fault.
+		 */
+		*(volatile uint8_t *)(&reg->STATUS) = 0x00;
+		*(volatile uint8_t *)(&reg->STATUS) = 0x40;
+		*(volatile uint8_t *)(&reg->STATUS) = 0xC0;
+		*(volatile uint8_t *)(&reg->STATUS) = 0x80;
+	}
+}
+
+static int nxp_irtc_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+
+	if (!timeptr || !is_rtc_time_in_bounds(timeptr, 0)) {
+		return -EINVAL;
+	}
+
+	int calc_year = timeptr->tm_year - RTC_NXP_IRTC_YEAR_OFFSET;
+	/* The IRTC Month Index starts at 1 instead of 0 */
+	int calc_month = timeptr->tm_mon + 1;
+
+	uint32_t key = irq_lock();
```

Comment:
Same note here, use a binary semaphore for locking (or don't lock at all), it looks like setting the time may take a while, and we don't want to block every other thread while doing so :)

Reply:
Should this be changed, I have irq_lock for the same reason as the other irq_lock, these registers could
lock up during a context switch

Reply:
ok, its fine to use irq_lock here, k_spinlock would be the portable way of preventing context switching (it essentially just calls irq_lock() for most platforms)